File: 4th semester/Algorithm Analysis/Merge_Sort.cpp
#include <bits/stdc++.h>
using namespace std; 

vector<int>merge_sort(int n, vector<int>&arr){
    if(n > 1){
        int v1 = n / 2;
        int v2 = n - v1;
        vector<int>left(v1), right(v2);
        for(int i = 0; i < v1; i++){
            left[i]=arr[i];
        }
        for(int i = v1; i < n; i++){
            right[i - v1] = arr[i];
        }
        merge_sort(v1, left);
        merge_sort(v2, right);
        int i = 0, j = 0, k = 0;
        while(i < v1 and j < v2){
            if(left[i] < right[j]){
                arr[k] = left[i];
                i++; k++;
            }
            else{
                arr[k] = right[j];
                j++; k++;
            }
        }
        while(i < v1){
            arr[k] = left[i];
            i++; k++;
        }
        while(j < v2){
            arr[k] = right[j];
            j++; k++;
        }
    }
    return arr;
}

int main(){
    cout << "Enter the size of array: ";
    int n; cin >> n;
    cout << "Enter the element of array: ";
    vector<int>arr(n);
    for(int i = 0; i < n; i++){
        cin >> arr[i];
    }
    cout << "After sorting Merge sort, the array is: ";
    for(auto i: merge_sort(n, arr)){
        cout << i << ' ';
    }
    cout << '\n';
    return 0;
}

File: 4th semester/Algorithm Analysis/Merge_Sort.py
def merge(arr:list)->list:
    if len(arr)>1:
        left,right=arr[:len(arr)//2],arr[len(arr)//2:]
        merge(left)
        merge(right)
        i,j,k=0,0,0
        while i<len(left) and j<len(right):
            if left[i]<right[j]:
                arr[k]=left[i]
                i+=1
            else:
                arr[k]=right[j]
                j+=1
            k+=1
        while i<len(left):
            arr[k]=left[i]
            k+=1
            i+=1
        while j<len(right):
            arr[k]=right[j]
            k+=1
            j+=1
    return arr

if __name__=="__main__":
    n=int(input("Enter the size of array: "))
    arr=[*map(int,input("Enter the element of array: ").split())][:n]
    print("The sorted array :",*merge(arr))
